src/Core/Grid/Query/TaxRuleQueryBuilder.php

```diff
@@ -0,0 +1,156 @@
+    public function __construct(
+        Connection $connection,
+        $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator,
+        int $employeeIdLang
+    ) {
+        parent::__construct($connection, $dbPrefix);
+
+        $this->searchCriteriaApplicator = $searchCriteriaApplicator;
+        $this->employeeIdLang = $employeeIdLang;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSearchQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getQueryBuilder($searchCriteria->getFilters());
+
+        $qb
+            ->select([
+                'tr.`id_tax_rule`',
+                'tr.`description`',
+                'cl.`name` AS country_name',
+                'IFNULL(s.`name`, \'--\') AS state_name',
+                'CASE '
+                    . ' WHEN CONCAT_WS(\' - \', tr.`zipcode_from`, tr.`zipcode_to`) = \'0 - 0\''
+                    . ' THEN \'--\' ELSE CONCAT_WS(\' - \', tr.`zipcode_from`, tr.`zipcode_to`)'
+                . ' END AS zipcode',
+                'tr.behavior',
+                't.rate',
+            ])
+        ;
+
+        $this->searchCriteriaApplicator
+            ->applyPagination($searchCriteria, $qb)
+            ->applySorting($searchCriteria, $qb)
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        return $this
+            ->getQueryBuilder($searchCriteria->getFilters())
+            ->select('COUNT(DISTINCT tr.`id_tax_rule`)');
+    }
+
+    /**
+     * Gets query builder with the common sql used for displaying tax rule groups list and applying filter actions.
+     *
+     * @param array $filters
+     *
+     * @return QueryBuilder
+     */
+    private function getQueryBuilder(array $filters): QueryBuilder
+    {
+        $qb = $this->connection->createQueryBuilder();
+
+        $qb
+            ->from($this->dbPrefix . 'tax_rule', 'tr');
+
+        $qb->leftJoin(
+            'tr',
+            $this->dbPrefix . 'country',
+            'c',
+            'tr.`id_country` = c.`id_country`'
+        );
+
+        $qb->leftJoin(
+            'tr',
+            $this->dbPrefix . 'country_lang',
+            'cl',
+            'tr.`id_country` = cl.`id_country` AND cl.`id_lang` = :idLang '
+        );
+
+        $qb->leftJoin(
+            'tr',
+            $this->dbPrefix . 'state',
+            's',
+            'tr.`id_country` = s.`id_country` AND tr.`id_state` = s.`id_state`'
+        );
+
+        $qb->leftJoin(
+            'tr',
+            $this->dbPrefix . 'tax',
+            't',
+            'tr.`id_tax` = t.`id_tax`'
+        );
+
+        $qb->andWhere('tr.`id_tax_rules_group` = :idTaxRulesGroup');
+
+        $qb
+            ->setParameter('idLang', $this->employeeIdLang)
+            ->setParameter('idTaxRulesGroup', $filters['taxRulesGroupId']);
+
+        return $qb;
```

Comment:
lots of unnecessary separationss. It could all be done in one like:
```
        return $this->connection->createQueryBuilder()
            ->from($this->dbPrefix . 'tax_rule', 'tr')
            ->leftJoin(
                'tr',
                $this->dbPrefix . 'country',
                'c',
                'tr.`id_country` = c.`id_country`'
            )
            ->leftJoin(
                'tr',
                $this->dbPrefix . 'country_lang',
                'cl',
                'tr.`id_country` = cl.`id_country` AND cl.`id_lang` = :idLang '
            )
            .... etc
```